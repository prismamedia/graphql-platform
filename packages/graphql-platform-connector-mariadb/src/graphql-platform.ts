import {
  BaseContext as CoreBaseContext,
  Context as CoreContext,
  CustomContext,
  CustomOperationConfig as CoreCustomOperationConfig,
  GraphQLPlatform as CoreGraphQLPlatform,
  GraphQLPlatformConfig as CoreGraphQLPlatformConfig,
} from '@prismamedia/graphql-platform-core';
import { Maybe, POJO } from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import {
  Connector,
  ConnectorConfig,
  ConnectorRequest,
} from './graphql-platform/connector';
import { ResourceConfig } from './graphql-platform/resource';

export * from './graphql-platform/connector';
export * from './graphql-platform/resource';

/** "Context" provided by the GraphQL Platform */
export type BaseContext = CoreBaseContext & {
  /**
   * Use the "Connector" at your own risk as you can do whatever you want with the database
   */
  connector: Connector;

  /**
   * A new "ConnectorRequest" is created for every GraphQL execution's context, it is a convenient place
   * to share a state/cache between all the resolvers of the same request
   */
  connectorRequest: ConnectorRequest;
};

export type AnyBaseContext = BaseContext;

export type Context<TCustomContext extends CustomContext = {}> = CoreContext<
  TCustomContext,
  BaseContext
>;

export type CustomOperationConfig<
  TArgs extends POJO = any,
  TCustomContext extends CustomContext = {}
> = CoreCustomOperationConfig<TArgs, TCustomContext, BaseContext>;

export interface GraphQLPlatformConfig<
  TContextParams extends POJO = any,
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext
> extends CoreGraphQLPlatformConfig<
    TContextParams,
    TCustomContext,
    TBaseContext,
    ResourceConfig<TCustomContext, TBaseContext>
  > {
  connector?: Maybe<ConnectorConfig>;
}

export class GraphQLPlatform<
  TContextParams extends POJO = any,
  TCustomContext extends CustomContext = {},
  TBaseContext extends AnyBaseContext = BaseContext
> extends CoreGraphQLPlatform<
  TContextParams,
  TCustomContext,
  TBaseContext,
  GraphQLPlatformConfig<TContextParams, TCustomContext, TBaseContext>
> {
  @Memoize()
  public getConnector(): Connector {
    return new Connector(this.config.connector, this);
  }

  public async getBaseContext(): Promise<TBaseContext> {
    return {
      ...(await super.getBaseContext()),
      connector: this.getConnector(),
      connectorRequest: this.getConnector().newRequest(),
    };
  }
}
