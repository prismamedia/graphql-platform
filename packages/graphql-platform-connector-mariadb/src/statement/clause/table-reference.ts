import type * as core from '@prismamedia/graphql-platform';
import type { Table } from '../../schema.js';

export abstract class AbstractTableReference {
  public abstract readonly alias: string;

  /**
   * The depth of a node is the number of edges from the node to the tree's root node.
   * A root node will have a depth of 0.
   */
  public abstract readonly depth: number;

  public readonly children = new Map<JoinTable['alias'], JoinTable>();

  public constructor(public readonly table: Table) {}

  /**
   * The height of a node is the number of edges on the longest path from the node to a leaf.
   * A leaf node will have a height of 0.
   */
  public get height(): number {
    return this.children.size
      ? Math.max(
          ...Array.from(this.children.values(), (child) => child.height),
        ) + 1
      : 0;
  }

  public join(edge: core.Edge | core.ReverseEdge, key?: string): JoinTable {
    const joinTable = new JoinTable(
      this,
      this.table.schema.getTableByNode(edge.head),
    );

    const maybeExistingJoinTable = this.children.get(joinTable.alias);

    // We join a table only once per its alias
    if (maybeExistingJoinTable) {
      return maybeExistingJoinTable;
    }

    this.children.set(joinTable.alias, joinTable);

    return joinTable;
  }
}

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class TableFactor extends AbstractTableReference {
  public override readonly alias: string;
  public override readonly depth: number = 0;

  public constructor(table: Table) {
    super(table);

    this.alias = table.name;
  }
}

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class JoinTable extends AbstractTableReference {
  public override readonly alias: string;
  public override readonly depth: number;

  public constructor(
    public readonly parent: TableReference,
    table: Table,
    /**
     * An optional key used to discriminate 2 joins, imagine the following "selection":
     *
     * articles {
     *  firstTags: tags(orderBy: [order_ASC], first: 5) { title }
     *  lastTags: tags(orderBy: [order_DESC], first: 5) { title }
     * }
     *
     * We need 2 joins on "tags", and we'll discriminate them using the keys "firstTags" and "lastTags"
     */
    key?: string,
  ) {
    super(table);

    this.alias = `${parent.alias}>${key || table.name}`;
    this.depth = parent.depth + 1;
  }
}

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export type TableReference = TableFactor | JoinTable;
