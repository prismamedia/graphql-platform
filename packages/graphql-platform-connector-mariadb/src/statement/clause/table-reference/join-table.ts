import type * as core from '@prismamedia/graphql-platform';
import { Memoize } from '@prismamedia/ts-memoize';
import { EOL } from 'node:os';
import { escapeIdentifier } from '../../../escape.js';
import { AbstractTableReference } from '../abstract-table-reference.js';
import type { TableReference } from '../table-reference.js';

export enum JoinTableKind {
  LEFT,
  INNER,
}

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class JoinTable extends AbstractTableReference {
  public override readonly alias: string;
  public override readonly depth: number;

  public constructor(
    public readonly parent: TableReference,
    public readonly edge: core.Edge | core.ReverseEdge,
    /**
     * An optional key used to discriminate 2 joins, imagine the following "selection":
     *
     * articles {
     *  firstTags: tags(orderBy: [order_ASC], first: 5) { title }
     *  lastTags: tags(orderBy: [order_DESC], first: 5) { title }
     * }
     *
     * We need 2 joins on "tags", and we'll discriminate them using the keys "firstTags" and "lastTags"
     */
    key?: string,
  ) {
    const head = parent.table.schema.getTableByNode(edge.head);
    super(head);

    this.alias = `${parent.alias}>${key || head.name}`;
    this.depth = parent.depth + 1;
  }

  public get kind(): JoinTableKind {
    return this.edge.isNullable() ? JoinTableKind.LEFT : JoinTableKind.INNER;
  }

  @Memoize()
  public get condition(): string {
    return 'test = test';
  }

  public override toString(): string {
    return [
      `${JoinTableKind[this.kind]} JOIN ${escapeIdentifier(
        this.table.qualifiedName,
      )} AS ${escapeIdentifier(this.alias)} ON ${this.condition}`,
      ...Array.from(this.children.values(), (joinTable) =>
        joinTable
          .toString()
          // Indent the children for better visibility
          .split(EOL)
          .map((line) => `  ${line}`)
          .join(EOL),
      ),
    ].join(EOL);
  }
}
