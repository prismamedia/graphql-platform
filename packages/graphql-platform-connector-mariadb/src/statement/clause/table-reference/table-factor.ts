import type { Table } from '../../../schema.js';
import { AbstractTableReference } from '../abstract-table-reference.js';

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class TableFactor extends AbstractTableReference {
  public override readonly alias: string;
  public override readonly depth: number = 0;

  public constructor(table: Table) {
    super(table);

    this.alias = table.name;
  }
}
