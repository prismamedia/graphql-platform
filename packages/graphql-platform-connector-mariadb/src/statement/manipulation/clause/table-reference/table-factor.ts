import type * as core from '@prismamedia/graphql-platform';
import { escapeIdentifier } from '../../../../escaping.js';
import type { Table } from '../../../../schema.js';
import { AbstractTableReference } from '../abstract-table-reference.js';

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class TableFactor extends AbstractTableReference {
  public override readonly alias: string;
  public override readonly depth: number = 0;

  public constructor(table: Table, context: core.OperationContext) {
    super(table, context);

    this.alias = table.name;
  }

  public override toString(): string {
    return [
      `${escapeIdentifier(this.table.name)}${
        this.alias === this.table.name
          ? ''
          : ` AS ${escapeIdentifier(this.alias)}`
      }`,
      ...Array.from(this.joinsByAlias.values(), (joinTable) =>
        joinTable.toString(),
      ),
    ].join(' ');
  }
}
