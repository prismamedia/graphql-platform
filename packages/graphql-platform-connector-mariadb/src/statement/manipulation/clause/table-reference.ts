import * as core from '@prismamedia/graphql-platform';
import * as utils from '@prismamedia/graphql-platform-utils';
import { MGetter, MMethod } from '@prismamedia/memoize';
import assert from 'node:assert';
import type { Merge } from 'type-fest';
import type { MariaDBBrokerSubscriptionsStateTable } from '../../../broker/table/subscriptions-state.js';
import { escapeIdentifier } from '../../../escaping.js';
import type { Column, Table } from '../../../schema.js';
import {
  AuthorizedTableCTE,
  AuthorizedTableDerivedTable,
  Subquery,
  type AbstractSelectOptions,
} from '../query/select.js';
import { AND, filterNode, type WhereCondition } from './where-condition.js';

export class InlineTableAuthorization {
  public readonly name: string;

  public constructor(
    public readonly tableReference: TableReference,
    public readonly filter: core.NodeFilter,
  ) {
    this.name = tableReference.table.name;
  }

  @MGetter
  public get condition(): WhereCondition {
    return filterNode(this.tableReference, this.filter);
  }
}

export type TableAuthorization =
  | AuthorizedTableCTE
  | AuthorizedTableDerivedTable
  | InlineTableAuthorization;

export abstract class AbstractTableReference {
  public abstract readonly root: TableFactor;
  public abstract readonly source: TableAuthorization | Table;
  public abstract readonly alias: string;

  public readonly joinsByAlias = new Map<JoinTable['alias'], JoinTable>();
  public readonly subqueries = new Set<TableFactor>();

  public constructor(
    public readonly table: Table,
    public readonly context?: core.OperationContext,
  ) {}

  public abstract toString(): string;

  public isRoot(): boolean {
    assert(isTableReference(this));

    return this.root === this;
  }

  public authorize(tableReference: TableReference): TableAuthorization | Table {
    const table = tableReference.table;

    const authorization = this.context?.getAuthorization(table.node);
    if (authorization) {
      if (authorization.isExecutableWithin(table.node.selection)) {
        return new InlineTableAuthorization(tableReference, authorization);
      } else if (this.root.useCommonTableExpression) {
        let authorizedTable = this.root.authorizedTables.get(table);

        if (!authorizedTable) {
          this.root.authorizedTables.set(
            table,
            (authorizedTable = new AuthorizedTableCTE(
              new TableFactor(table, { parent: tableReference }),
              authorization,
            )),
          );
        }

        return authorizedTable;
      } else {
        return new AuthorizedTableDerivedTable(
          new TableFactor(table, { parent: tableReference }),
          authorization,
        );
      }
    }

    return table;
  }

  public escapeColumnIdentifier(column: Column): string {
    assert(this.table.columns.includes(column));

    return escapeIdentifier(`${this.alias}.${column.name}`);
  }

  public escapeColumnIdentifierByLeaf(leaf: core.Leaf): string {
    return this.escapeColumnIdentifier(this.table.getColumnByLeaf(leaf));
  }

  public getJoinConditions(
    edgeOrReverseEdge: core.Edge | core.ReverseEdge,
    head: TableReference,
  ): Array<WhereCondition> {
    assert.strictEqual(edgeOrReverseEdge.tail, this.table.node);
    assert.strictEqual(edgeOrReverseEdge.head, head.table.node);

    return edgeOrReverseEdge instanceof core.Edge
      ? this.table
          .getForeignKeyByEdge(edgeOrReverseEdge)
          .getJoinConditions(this, head)
      : head.table
          .getForeignKeyByEdge(edgeOrReverseEdge.originalEdge)
          .getJoinConditions(head, this);
  }

  public join(
    edgeOrUniqueReverseEdge: core.Edge | core.UniqueReverseEdge,
  ): JoinTable {
    assert.strictEqual(edgeOrUniqueReverseEdge.tail, this.table.node);
    assert(isTableReference(this));

    const joinTable = new JoinTable(this, edgeOrUniqueReverseEdge);

    const maybeExistingJoinTable = this.joinsByAlias.get(joinTable.alias);

    // We join a table only once
    if (maybeExistingJoinTable) {
      return maybeExistingJoinTable;
    }

    this.joinsByAlias.set(joinTable.alias, joinTable);

    return joinTable;
  }

  public subquery(
    edgeOrReverseEdge: core.Edge | core.ReverseEdge,
    options?: Merge<AbstractSelectOptions, { where?: core.NodeFilter }>,
  ): Subquery {
    assert.strictEqual(edgeOrReverseEdge.tail, this.table.node);
    assert(isTableReference(this));

    const tableReference = new TableFactor(
      this.table.schema.getTableByNode(edgeOrReverseEdge.head),
      {
        context: this.context,
        alias: `${this.alias}>${edgeOrReverseEdge.name}`,
        parent: this,
      },
    );

    this.subqueries.add(tableReference);

    const joinConditions = this.getJoinConditions(
      edgeOrReverseEdge,
      tableReference,
    );

    const where = options?.where
      ? filterNode(tableReference, options.where)
      : undefined;

    return new Subquery(tableReference, {
      ...options,
      where: AND([...joinConditions, where], false),
      limit:
        options?.limit ??
        (edgeOrReverseEdge instanceof core.MultipleReverseEdge ? undefined : 1),
    });
  }
}

export interface TableFactorOptions {
  readonly context?: core.OperationContext;
  readonly alias?: string;
  readonly parent?: TableReference;
  readonly useCommonTableExpression?: boolean;
}

export class TableFactor extends AbstractTableReference {
  public readonly useCommonTableExpression: boolean;

  public override readonly root: TableFactor;
  public override readonly alias: string;
  public override readonly source: TableAuthorization | Table;

  public readonly authorizedTables = new Map<Table, AuthorizedTableCTE>();

  #joinSubscriptionsStateTable?: JoinSubscriptionsStateTable;

  public constructor(table: Table, options?: TableFactorOptions) {
    super(table, options?.context);

    this.useCommonTableExpression = utils.getOptionalFlag(
      options?.useCommonTableExpression,
      false,
    );

    this.root = options?.parent?.root ?? this;
    this.alias = options?.alias ?? table.name;
    this.source = !this.isRoot() && this.context ? this.authorize(this) : table;
  }

  public joinSubscriptionsState(
    subscriptionId: core.ChangesSubscriptionStream['id'],
  ): JoinSubscriptionsStateTable {
    assert(this.isRoot());

    let joinSubscriptionsStateTable = this.#joinSubscriptionsStateTable;
    if (joinSubscriptionsStateTable) {
      assert.strictEqual(
        joinSubscriptionsStateTable.subscriptionId,
        subscriptionId,
      );

      return joinSubscriptionsStateTable;
    }

    assert(
      this.table.subscriptionsStateTable,
      `The table "${this.table.name}" has no "subscriptions' state" table`,
    );

    return (this.#joinSubscriptionsStateTable ??=
      new JoinSubscriptionsStateTable(
        this,
        this.table.subscriptionsStateTable,
        subscriptionId,
      ));
  }

  @MMethod()
  public override toString(): string {
    return [
      this.source instanceof AuthorizedTableDerivedTable
        ? `${this.source} AS ${escapeIdentifier(this.alias)}`
        : `${escapeIdentifier(this.source.name)}${
            this.alias === this.source.name
              ? ''
              : ` AS ${escapeIdentifier(this.alias)}`
          }`,
      ...Array.from(this.joinsByAlias.values(), String),
      this.#joinSubscriptionsStateTable?.toString(),
    ]
      .filter(Boolean)
      .join(' ');
  }
}

export enum JoinTableKind {
  LEFT,
  INNER,
}

export class JoinTable extends AbstractTableReference {
  public override readonly root: TableFactor;
  public override readonly alias: string;
  public override readonly source: TableAuthorization | Table;

  public constructor(
    public readonly parent: TableReference,
    public readonly edgeOrUniqueReverseEdge: core.Edge | core.UniqueReverseEdge,
  ) {
    assert(
      edgeOrUniqueReverseEdge instanceof core.Edge ||
        edgeOrUniqueReverseEdge instanceof core.UniqueReverseEdge,
    );
    assert.strictEqual(edgeOrUniqueReverseEdge.tail, parent.table.node);

    super(
      parent.table.schema.getTableByNode(edgeOrUniqueReverseEdge.head),
      parent.context,
    );

    this.root = parent.root;
    this.alias = `${parent.alias}>${edgeOrUniqueReverseEdge.name}`;
    this.source = this.context ? this.authorize(this) : this.table;
  }

  @MGetter
  public get kind(): JoinTableKind {
    return (this.parent instanceof JoinTable &&
      this.parent.kind === JoinTableKind.LEFT) ||
      this.edgeOrUniqueReverseEdge.isNullable()
      ? JoinTableKind.LEFT
      : JoinTableKind.INNER;
  }

  @MGetter
  public get condition(): WhereCondition {
    const joinConditions = this.parent.getJoinConditions(
      this.edgeOrUniqueReverseEdge,
      this,
    );

    return AND(
      this.source instanceof InlineTableAuthorization
        ? [...joinConditions, this.source.condition]
        : joinConditions,
      false,
    );
  }

  @MMethod()
  public override toString(): string {
    return [
      `${JoinTableKind[this.kind]} JOIN ${
        this.source instanceof AuthorizedTableDerivedTable
          ? this.source
          : escapeIdentifier(this.source.name)
      } AS ${escapeIdentifier(this.alias)} ON ${this.condition}`,
      ...Array.from(this.joinsByAlias.values(), String),
    ].join(' ');
  }
}

export class JoinSubscriptionsStateTable {
  public readonly alias: string;

  public constructor(
    public readonly parent: TableReference,
    public readonly table: MariaDBBrokerSubscriptionsStateTable,
    public readonly subscriptionId: core.ChangesSubscriptionStream['id'],
  ) {
    this.alias = `${parent.alias}>${table.name}`;
  }

  public toString(): string {
    const subscriptionIdColumn = this.table.getColumnByName('subscriptionId');

    return `${JoinTableKind[JoinTableKind.LEFT]} JOIN ${escapeIdentifier(
      this.table.name,
    )} AS ${escapeIdentifier(this.alias)} ON ${AND([
      ...this.table.references.map(
        ({ source, target }) =>
          `${escapeIdentifier(`${this.alias}.${source.name}`)} = ${this.parent.escapeColumnIdentifier(target)}`,
      ),
      `${escapeIdentifier(`${this.alias}.${subscriptionIdColumn.name}`)} = ${subscriptionIdColumn.dataType.serialize(this.subscriptionId)}`,
    ])}`;
  }

  public escapeColumnIdentifier(columnOrName: Column | Column['name']): string {
    const column =
      typeof columnOrName === 'string'
        ? this.table.getColumnByName(columnOrName)
        : columnOrName;
    assert(column);

    return escapeIdentifier(`${this.alias}.${column.name}`);
  }
}

export type TableReference = TableFactor | JoinTable;

export function isTableReference(
  maybeTableReference: unknown,
): maybeTableReference is TableReference {
  return (
    maybeTableReference instanceof TableFactor ||
    maybeTableReference instanceof JoinTable
  );
}
