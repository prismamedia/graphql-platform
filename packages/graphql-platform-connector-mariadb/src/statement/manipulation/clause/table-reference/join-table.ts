import * as core from '@prismamedia/graphql-platform';
import { Memoize } from '@prismamedia/ts-memoize';
import { escapeIdentifier } from '../../../../escaping.js';
import { AbstractTableReference } from '../abstract-table-reference.js';
import type { TableReference } from '../table-reference.js';
import { filterNode } from '../where-condition.js';

export enum JoinTableKind {
  LEFT,
  INNER,
}

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class JoinTable extends AbstractTableReference {
  protected readonly authorization?: core.NodeFilter;
  public override readonly alias: string;
  public override readonly depth: number;

  public constructor(
    public readonly parent: TableReference,
    public readonly edge: core.Edge | core.ReverseEdgeUnique,
  ) {
    const head = parent.table.schema.getTableByNode(edge.head);
    super(head, parent.context);

    this.authorization = parent.context.getAuthorization(edge.head);
    this.alias = `${parent.alias}>${head.name}`;
    this.depth = parent.depth + 1;
  }

  public get kind(): JoinTableKind {
    return this.edge.isNullable() ? JoinTableKind.LEFT : JoinTableKind.INNER;
  }

  @Memoize()
  public get condition(): string {
    return [
      ...(this.edge instanceof core.Edge
        ? this.parent.table
            .getForeignKeyByEdge(this.edge)
            .columns.map(
              (column) =>
                `${this.parent.getEscapedColumnIdentifier(
                  column,
                )} <=> ${this.getEscapedColumnIdentifier(
                  column.referencedColumn,
                )}`,
            )
        : this.table
            .getForeignKeyByEdge(this.edge.originalEdge)
            .columns.map(
              (column) =>
                `${this.parent.getEscapedColumnIdentifier(
                  column.referencedColumn,
                )} <=> ${this.getEscapedColumnIdentifier(column)}`,
            )),
      this.authorization ? filterNode(this, this.authorization) : undefined,
    ]
      .filter(Boolean)
      .join(' AND ');
  }

  public override toString(): string {
    return [
      `${JoinTableKind[this.kind]} JOIN ${escapeIdentifier(
        this.table.name,
      )} AS ${escapeIdentifier(this.alias)} ON ${this.condition}`,
      ...Array.from(this.joinsByAlias.values(), (joinTable) =>
        joinTable.toString(),
      ),
    ].join(' ');
  }
}
