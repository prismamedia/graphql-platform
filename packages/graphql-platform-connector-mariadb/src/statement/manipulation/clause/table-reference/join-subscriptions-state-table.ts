import type * as core from '@prismamedia/graphql-platform';
import assert from 'node:assert';
import type { Column } from '../../../../broker/abstract-table.js';
import type { MariaDBBrokerSubscriptionsStateTable } from '../../../../broker/table/subscriptions-state.js';
import { escapeIdentifier } from '../../../../escaping.js';
import type { TableReference } from '../table-reference.js';
import { AND } from '../where-condition.js';
import { JoinTableKind } from './join-table.js';

/**
 * @see https://mariadb.com/kb/en/join-syntax/
 */
export class JoinSubscriptionsStateTable {
  public readonly alias: string;
  public readonly depth: number;

  public constructor(
    public readonly parent: TableReference,
    public readonly table: MariaDBBrokerSubscriptionsStateTable,
    public readonly subscriptionId: core.ChangesSubscriptionStream['id'],
  ) {
    this.alias = `${parent.alias}>${table.name}`;
    this.depth = parent.depth + 1;
  }

  public toString(): string {
    const subscriptionIdColumn = this.table.getColumnByName('subscriptionId');

    return `${JoinTableKind[JoinTableKind.LEFT]} JOIN ${escapeIdentifier(
      this.table.name,
    )} AS ${escapeIdentifier(this.alias)} ON ${AND([
      ...this.table.references.map(
        ({ source, target }) =>
          `${escapeIdentifier(`${this.alias}.${source.name}`)} = ${this.parent.escapeColumnIdentifier(target)}`,
      ),
      `${escapeIdentifier(`${this.alias}.${subscriptionIdColumn.name}`)} = ${subscriptionIdColumn.dataType.serialize(this.subscriptionId)}`,
    ])}`;
  }

  public escapeColumnIdentifier(columnOrName: Column | Column['name']): string {
    const column =
      typeof columnOrName === 'string'
        ? this.table.getColumnByName(columnOrName)
        : columnOrName;
    assert(column);

    return escapeIdentifier(`${this.alias}.${column.name}`);
  }
}
