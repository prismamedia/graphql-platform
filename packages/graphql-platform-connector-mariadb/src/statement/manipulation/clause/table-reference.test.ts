import { afterAll, beforeAll, describe, expect, it } from '@jest/globals';
import { OperationContext } from '@prismamedia/graphql-platform';
import {
  myAdminContext,
  myJournalistContext,
  myUserContext,
  myVisitorContext,
} from '@prismamedia/graphql-platform/__tests__/config.js';
import { format } from '@sqltools/formatter';
import { createMyGP, type MyGP } from '../../../__tests__/config.js';
import { TableFactor } from './table-reference.js';

describe('Table reference', () => {
  let gp: MyGP;

  beforeAll(async () => {
    gp = createMyGP('connector_mariadb_table_reference_clause');

    await gp.connector.setup();
  });

  afterAll(() => gp.connector.teardown());

  it.each([
    myVisitorContext,
    myUserContext,
    myJournalistContext,
    myAdminContext,
  ])('generates valid SQL with authorization', (requestContext) => {
    const ArticleNode = gp.getNodeByName('Article');

    const ArticleCategoryEdge = ArticleNode.getEdgeByName('category');

    const ArticleCreatedByEdge = ArticleNode.getEdgeByName('createdBy');

    const UserNode = ArticleCreatedByEdge.head;
    const UserProfileReverseEdge =
      UserNode.getUniqueReverseEdgeByName('profile');

    const from = new TableFactor(
      gp.connector.schema.getTableByNode(ArticleNode),
      new OperationContext(gp, requestContext),
    );

    from.join(ArticleCategoryEdge);
    from.join(ArticleCreatedByEdge).join(UserProfileReverseEdge);

    expect(format(`SELECT * FROM ${from}`)).toMatchSnapshot();
  });

  it('generates valid SQL', () => {
    const ArticleNode = gp.getNodeByName('Article');
    const ArticleCategoryEdge = ArticleNode.getEdgeByName('category');
    const CategoryNode = ArticleCategoryEdge.head;
    const CategoryParentEdge = CategoryNode.getEdgeByName('parent');

    const from = new TableFactor(
      gp.connector.schema.getTableByNode(ArticleNode),
      new OperationContext(gp, myAdminContext),
    );

    from
      .join(ArticleCategoryEdge)
      .join(CategoryParentEdge)
      .join(CategoryParentEdge)
      .join(CategoryParentEdge);

    const ArticleCreatedByEdge = ArticleNode.getEdgeByName('createdBy');
    const UserNode = ArticleCreatedByEdge.head;
    const UserProfileReverseEdge =
      UserNode.getUniqueReverseEdgeByName('profile');

    from.join(ArticleCreatedByEdge).join(UserProfileReverseEdge);

    expect(format(`SELECT * FROM ${from}`)).toMatchInlineSnapshot(`
      "SELECT *
      FROM \`articles\`
        LEFT JOIN \`categories\` AS \`articles>category\` ON \`articles\`.\`category_private_id\` <= > \`articles>category\`.\`private_id\`
        LEFT JOIN \`categories\` AS \`articles>category>parent\` ON \`articles>category\`.\`parent_private_id\` <= > \`articles>category>parent\`.\`private_id\`
        LEFT JOIN \`categories\` AS \`articles>category>parent>parent\` ON \`articles>category>parent\`.\`parent_private_id\` <= > \`articles>category>parent>parent\`.\`private_id\`
        LEFT JOIN \`categories\` AS \`articles>category>parent>parent>parent\` ON \`articles>category>parent>parent\`.\`parent_private_id\` <= > \`articles>category>parent>parent>parent\`.\`private_id\`
        LEFT JOIN \`users\` AS \`articles>createdBy\` ON \`articles\`.\`created_by_id\` <= > \`articles>createdBy\`.\`id\`
        LEFT JOIN \`user_profiles\` AS \`articles>createdBy>profile\` ON \`articles>createdBy\`.\`id\` = \`articles>createdBy>profile\`.\`theUserId\`"
    `);
  });
});
