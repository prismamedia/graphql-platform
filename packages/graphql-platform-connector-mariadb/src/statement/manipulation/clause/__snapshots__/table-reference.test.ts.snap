// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table reference generates valid SQL with authorization 1`] = `
"SELECT *
FROM \`articles\`
  LEFT JOIN \`categories\` AS \`articles>categories\` ON \`articles\`.\`category_privateId\` <= > \`articles>categories\`.\`privateId\`
  LEFT JOIN \`users\` AS \`articles>users\` ON \`articles\`.\`created_by_id\` <= > \`articles>users\`.\`id\`
  AND FALSE
  LEFT JOIN \`user_profiles\` AS \`articles>users>user_profiles\` ON \`articles>users\`.\`id\` <= > \`articles>users>user_profiles\`.\`theUserId\`"
`;

exports[`Table reference generates valid SQL with authorization 2`] = `
"SELECT *
FROM \`articles\`
  LEFT JOIN \`categories\` AS \`articles>categories\` ON \`articles\`.\`category_privateId\` <= > \`articles>categories\`.\`privateId\`
  LEFT JOIN \`users\` AS \`articles>users\` ON \`articles\`.\`created_by_id\` <= > \`articles>users\`.\`id\`
  LEFT JOIN \`user_profiles\` AS \`articles>users>user_profiles\` ON \`articles>users\`.\`id\` <= > \`articles>users>user_profiles\`.\`theUserId\`"
`;

exports[`Table reference generates valid SQL with authorization 3`] = `
"SELECT *
FROM \`articles\`
  LEFT JOIN \`categories\` AS \`articles>categories\` ON \`articles\`.\`category_privateId\` <= > \`articles>categories\`.\`privateId\`
  LEFT JOIN \`users\` AS \`articles>users\` ON \`articles\`.\`created_by_id\` <= > \`articles>users\`.\`id\`
  LEFT JOIN \`user_profiles\` AS \`articles>users>user_profiles\` ON \`articles>users\`.\`id\` <= > \`articles>users>user_profiles\`.\`theUserId\`"
`;

exports[`Table reference generates valid SQL with authorization 4`] = `
"SELECT *
FROM \`articles\`
  LEFT JOIN \`categories\` AS \`articles>categories\` ON \`articles\`.\`category_privateId\` <= > \`articles>categories\`.\`privateId\`
  LEFT JOIN \`users\` AS \`articles>users\` ON \`articles\`.\`created_by_id\` <= > \`articles>users\`.\`id\`
  LEFT JOIN \`user_profiles\` AS \`articles>users>user_profiles\` ON \`articles>users\`.\`id\` <= > \`articles>users>user_profiles\`.\`theUserId\`"
`;
